crates/noirc_evaluator/src/ssa/conditional.rs

```diff
@@ -0,0 +1,485 @@
+            );
+            DecisionTree::new_instruction_after_phi(
+                ctx,
+                block_id,
+                BinaryOp::Mul,
+                pvalue,
+                not_condition,
+                ObjectType::Boolean,
+            )
+        };
+        self[assumption_id].value = Some(ins);
+        ins
+    }
+
+    pub fn make_decision_tree(&mut self, ctx: &mut SsaContext) {
+        let mut join_to_process = Vec::new();
+        let mut join_processed = Vec::new();
+        self.decision_tree(
+            ctx,
+            self.root,
+            ctx.first_block,
+            &mut join_to_process,
+            &mut join_processed,
+        );
+    }
+
+    pub fn decision_tree(
+        &mut self,
+        ctx: &mut SsaContext,
+        current_assumption: AssumptionId,
+        block_id: BlockId,
+        join_to_process: &mut Vec<BlockId>,
+        join_processed: &mut Vec<BlockId>,
+    ) {
+        let assumption = &self[current_assumption];
+        let block = &ctx[block_id];
+        let mut block_assumption = current_assumption;
+        let mut if_assumption = None;
+        let mut parent = AssumptionId::dummy();
+        let mut sibling = true;
+        if join_processed.contains(&block_id) {
+            return;
+        }
+        // is it an exit block?
+        if join_to_process.contains(&block_id) {
+            debug_assert!(block_id == *join_to_process.last().unwrap());
+            block_assumption = assumption.parent;
+            join_to_process.pop();
+            join_processed.push(block_id);
+        }
+        // is it an IF block?
+        if let Some(ins) = ctx.try_get_instruction(*block.instructions.last().unwrap()) {
+            if !block.is_join() && ins.operation.opcode() == super::node::Opcode::Jeq {
+                //add a new assuption for the IF
+                if assumption.parent == AssumptionId::dummy() {
+                    //Root assumption
+                    parent = current_assumption;
+                    sibling = true;
+                } else {
+                    parent = assumption.parent;
+                    sibling = self[assumption.parent].val_true.contains(&current_assumption);
+                };
+                let mut if_decision = Assumption::new(parent);
+                if let Operation::Jeq(condition, _) = ins.operation {
+                    if_decision.condition = condition;
+                } else {
+                    unreachable!();
+                }
+
+                //find exit node:
+                let exit = block::find_join(ctx, block.left.unwrap(), block.right.unwrap());
+                assert!(ctx[exit].kind == BlockType::IfJoin); //todo debug_assert
+                if_decision.entry_block = block_id;
+                if_decision.exit_block = exit;
+                if_assumption = Some(if_decision);
+                join_to_process.push(exit);
+            }
+        }
+        //let's mutate
```

Comment:
This comment is not helpful, lets expand it or remove it

Reply:
Done